api/saml/metadata/model.py

```diff
@@ -1235,74 +1238,154 @@ def decode(self, raw_subject, _w=WHITESPACE.match):
-           [https://docs.oasis-open.org/security/saml-subject-id-attr/v1.0/saml-subject-id-attr-v1.0.html],
-           one of which ("urn:oasis:names:tc:SAML:attribute:pairwise-id") is a direct replacement for this identifier
-           with a simpler syntax and safer comparison rules.
-           Existing use of this attribute in SAML 1.1 or SAML 2.0 should be phased out
-           in favor of the new Subject Identifier attributes."
+            if match:
+                patron_id = match.group(self.PATRON_ID_REGULAR_EXPRESSION_NAMED_GROUP)
+        else:
+            patron_id = patron_id_candidate
 
-        3. uid
-           http://oid-info.com/get/0.9.2342.19200300.100.1.1
+        return patron_id
 
-           See IETF RFC 4519.
-           IETF RFC 1274 uses the identifier "userid".
+    def extract(self, subject):
+        """Extract a unique patron ID from the SAML subject.
 
-        4. Name ID
-           The extractor fetches the first name ID it could find as a last resort which may no be correct.
-           It might be better to fetch only persistent name IDs.
+        :param subject: SAML subject
+        :type subject: SAMLSubject
 
-        Also, please note that eduPersonTargetedID attribute and name IDs should be phased out and replaced with
-        the pairwise-id attribute from the OASIS SAML 2.0 SubjectID Attributes Profile.
-        However, it's not yet supported by most of the IdPs
+        :return: Unique ID
+        :rtype: string
+
+        :return: Unique patron ID if any
+        :rtype: Optional[str]
         """
-        unique_id_attributes = [
-            SAMLAttributeType.eduPersonUniqueId,
-            SAMLAttributeType.eduPersonTargetedID,
-            SAMLAttributeType.uid,
-        ]
+        self._logger.info(
+            u"Trying to extract a unique patron ID from {0}".format(
+                six.ensure_text(repr(subject))
+            )
+        )
+
+        patron_id = None
 
         if subject.attribute_statement:
-            for unique_id_attribute in unique_id_attributes:
-                if unique_id_attribute.name in subject.attribute_statement.attributes:
-                    unique_id_attribute = subject.attribute_statement.attributes[
-                        unique_id_attribute.name
+            for patron_id_attribute in self._patron_id_attributes:
+                if patron_id_attribute in subject.attribute_statement.attributes:
+                    patron_id_attribute = subject.attribute_statement.attributes[
+                        patron_id_attribute
                     ]
 
-                    return unique_id_attribute.values[0]
+                    # NOTE: It takes the first value.
+                    patron_id_candidate = patron_id_attribute.values[0]
+                    patron_id = self._extract_patron_id(patron_id_candidate)
+
+                    if patron_id is not None:
+                        break
+
+        if self._use_name_id:
```

Comment:
Here it looks like the presence of NameID could override a previously set patron_id.

Reply:
You're absolutely right, fixed the issue and added additional tests.